### PR TITLE
handle Ctrl+C+C can not show popup window in many Qt programs

### LINKED ISSUE #1420: ctrl+c+c popup doesn't work on Ubuntu 21.10 with firefox, evince, sublime text
ctrl+c+c clipboard popup doesn't work on Ubuntu 21.10 (Gnome Shell 40.5) with firefox, evince, sublime text,... but works with chromium, chrome, VS code,...

### PR DESCRIPTION
**Pull request by wisherhxl**
Ctrl+C+C have some strange problems in different systems and in different Qt applications.
for example, in my current laptop, if NumLock is on, In QtCreator, Ctrl+C+C can not show popup window, and when CapsLock is on, the problem vanished.

I guess maybe that's because when the popup is triggered, the other Qt program have not finished the setText procedure.

So I connectted the dataChanged() signal to the translateWord method and add some key to make it pop out only when triggered.

tested under Ubuntu, and it worked well.

Resolves #1420, and maybe #650, #858, #1290 too.

### CODE REVIEW

File: scanpopup.cc [hunk 1]
```diff
@@ -270,6 +270,9 @@ ScanPopup::ScanPopup( QWidget * parent,
   connect( definition, SIGNAL( titleChanged(  ArticleView *, QString const & ) ),
            this, SLOT( titleChanged(  ArticleView *, QString const & ) ) );
 
+  connect( QApplication::clipboard(), SIGNAL( dataChanged() ),
+           this, SLOT( handleTranslateWordFromClipboard() ) );
```
Comment (nonwill) [inline scanpopup.cc:274]: 这个方案通用吗？
>void QClipboard::dataChanged()
This signal is emitted when the clipboard data is changed.
>
>On macOS and with Qt version 4.3 or higher, clipboard changes made by other applications will only be detected when the application is activated.

Comment (vedgy) [reply scanpopup.cc:274]: This must be breaking the GD popup when the contents of the selection buffer changes (see the connection under `#ifdef HAVE_X11` below).

**Review by vedgy — CHANGES_REQUESTED** [review]
Perhaps a much simpler solution would work - just turn the following connection into a `Qt::QueuedConnection`:
```cpp
  connect( QApplication::clipboard(), SIGNAL( changed( QClipboard::Mode ) ),
           this, SLOT( clipboardChanged( QClipboard::Mode ) ) );
```

File: scanpopup.hh [hunk 2]
```diff
@@ -236,7 +242,6 @@ private slots:
 
 #ifdef HAVE_X11
   void delayShow();
-#endif
+#endif 
 };
-
```
Comment (vedgy) [inline scanpopup.hh:241]: Revert these unrelated formatting changes.

File: scanpopup.hh [hunk 3]
```diff
@@ -125,6 +124,10 @@ private:
 
   Config::Class & cfg;
   bool isScanningEnabled;
+  bool isTranslateWordFromClipboard = false;
+  QTimer translateWordFromClipboardDelayTimer;
+  QTimer translateWordFromClipboardClearTimer;
+  QClipboard::Mode TranslateWordFromClipboardMode = QClipboard::Clipboard;
```
Comment (vedgy) [inline scanpopup.hh:130]: Code style: variable names must start with a lowercase letter.

File: scanpopup.hh [hunk 4]
```diff
@@ -125,6 +124,10 @@ private:
 
   Config::Class & cfg;
   bool isScanningEnabled;
+  bool isTranslateWordFromClipboard = false;
```
Comment (vedgy) [inline scanpopup.hh:127]: GoldenDict aims to support building with C++98, so data members have to be initialized in the constructor.

File: scanpopup.cc (hunk 1 above)
Comment (wisherhxl) [reply scanpopup.cc:274]: The windows and macOS platforms do not have clipboard issues, so the `translateWordFromClipboardDelayTimer` will do the job. 

The  `QClipboard::dataChanged()` is used to handle the clipboard issue on linux with other Qt programs.

I've considered calling `handleTranslateWordFromClipboard()` directly, but using a timer seems better.

File: scanpopup.cc [hunk 5]
```diff
@@ -467,28 +482,49 @@ Qt::WindowFlags ScanPopup::unpinnedWindowFlags() const
...
+  if(!isTranslateWordFromClipboard) return;
+
   GD_DPRINTF( "translating from clipboard or selection\n" );
 
   QString subtype = "plain";
 
-  QString str = QApplication::clipboard()->text( subtype, m);
+  QString str = QApplication::clipboard()->text( subtype, TranslateWordFromClipboardMode );
+
+  if(!str.isEmpty()) {
```
Comment (vedgy) [inline scanpopup.cc:517]: Code style: don't attach brackets.

File: scanpopup.cc [hunk 6]
```diff
@@ -467,28 +482,49 @@ Qt::WindowFlags ScanPopup::unpinnedWindowFlags() const
...
 
 void ScanPopup::editGroupRequested()
 {
   emit editGroupRequested( ui.groupList->getCurrentGroup() );
 }
 
-void ScanPopup::translateWordFromClipboard(QClipboard::Mode m)
+void ScanPopup::handleTranslateWordFromClipboard()
 {
+  if(!isTranslateWordFromClipboard) return;
```
Comment (vedgy) [inline scanpopup.cc:509]: Code style:
* move `return;` to the next line;
* pad with spaces inside parentheses.

File: scanpopup.cc (hunk 1 above)
Comment (vedgy) [reply scanpopup.cc:274]: I mean the connection you added duplicates and reduces the functionality of the already existing connection just below, because the `dataChanged()` signal is a specialization of the already connected-to `changed( QClipboard::Mode )` signal!

**Review by vedgy — CHANGES_REQUESTED** [review]
I added some nitpicks, but I very much doubt your current approach is good. Try just adding `, Qt::QueuedConnection` to the existing connection instead as I suggested in the previous review.

**Comment by wisherhxl** [comment]
> Perhaps a much simpler solution would work - just turn the following connection into a `Qt::QueuedConnection`:
> 
> ```c++
>   connect( QApplication::clipboard(), SIGNAL( changed( QClipboard::Mode ) ),
>            this, SLOT( clipboardChanged( QClipboard::Mode ) ) );
> ```

**Comment by vedgy** [comment]
@wisherhxl, I think you completely misunderstand how clipboard and selection translation works in GoldenDict. Ctrl+C+C is only one of the ways to trigger scan popup on GNU/Linux. As a consequence, the current code must be breaking selection translation. Have you even tried using it? I mean: check the Scan Popup checkbox in the notification area (AKA systray) icon's context menu, then select a word in some application.

File: scanpopup.cc (hunk 1 above)
Comment (wisherhxl) [reply scanpopup.cc]: I guess `handleTranslateWordFromClipboard()` should have a higher priority, so I just put it into `ScanPopup::clipboardChanged( QClipboard::Mode m )`.

What do you think?

File: scanpopup.hh (hunk 2 above)
Comment (wisherhxl) [reply scanpopup.hh]: fixed.

File: scanpopup.hh (hunk 3 above)
Comment (wisherhxl) [reply scanpopup.hh]: fixed

File: scanpopup.hh (hunk 4 above)
Comment (wisherhxl) [reply scanpopup.hh]: fixed

File: scanpopup.cc (hunk 5 above)
Comment (wisherhxl) [reply scanpopup.cc]: fixed

File: scanpopup.cc (hunk 6 above)
Comment (wisherhxl) [reply scanpopup.cc]: fixed

**Comment by wisherhxl** [comment]
> @wisherhxl, I think you completely misunderstand how clipboard and selection translation works in GoldenDict. Ctrl+C+C is only one of the ways to trigger scan popup on GNU/Linux. As a consequence, the current code must be breaking selection translation. Have you even tried using it? I mean: check the Scan Popup checkbox in the notification area (AKA systray) icon's context menu, then select a word in some application.



> Contributor

Hi,  @vedgy . You are right, I have never used Scan Popup before, and new to Linux too. 
So I reconsidered the way to handle the issue, and hope the new solution could fix the Ctrl+C+C without breaking the Scan Popup.

Ctrl+C+C is an one time action triggered by user intentionally, so I think it has a higher priority than Scan Popup, which consistently pops out. So put it into `ScanPopup::clipboardChanged( QClipboard::Mode m )` with a higher priority should be ok.

And now, I've tested, and it worked well.

**Comment by nonwill** [comment]
Ctrl+C can cause unexpected problems on various systems(Windows/macOS/Linux) because it is a system wide feature and a lot of software is listening for its event. On Linux Scan Popup by select words is always the best way to us.

**Comment by nonwill** [comment]
>check the Scan Popup checkbox in the notification area (AKA systray) icon's context menu, then select a word in some application.

On macOS, it also has a more easy way to do this - by mouse hovering on text with a hotkey to show the Popup. But on Windows QClipboard has no select event for users to capture(no easy way to show the Popup by selecting words) , Ctrl+C(+C) often behaves badly, and hovering function need to hook to other applications system-widely, also lead to significant impact on system stability.

**Comment by wisherhxl** [comment]
@nonwill I agree, so the latest solution is to treat `Ctrl+C(+C)` as `Scan Popup`, and as I've tested, it works well.

**Comment by vedgy** [comment]
> And now, I've tested, and it worked well.

I don't know how you tested it, but the current code looks completely wrong. You clearly still don't understand the code related to `QClipboard::Mode`. Have you tried my suggestion? Instead of all the changes in this pull request, test the simple patch:

```diff
   connect( QApplication::clipboard(), SIGNAL( changed( QClipboard::Mode ) ),
-           this, SLOT( clipboardChanged( QClipboard::Mode ) ) );
+           this, SLOT( clipboardChanged( QClipboard::Mode ) ), Qt::QueuedConnection );
```

**Comment by wisherhxl** [comment]
@vedgy The new commit removed the connection and make `ScanPopup::clipboardChanged( QClipboard::Mode m )` handle it instead. so no duplicates esxists.
```diff
- connect( QApplication::clipboard(), SIGNAL( dataChanged() ),
-               this, SLOT( handleTranslateWordFromClipboard() ) );

void ScanPopup::clipboardChanged( QClipboard::Mode m )
{
+  if ( isTranslateWordFromClipboard && m == translateWordFromClipboardMode )
+    return handleTranslateWordFromClipboard();
```

If ScanPop is enabled, when the clipboard changed, `isTranslateWordFromClipboard` is false, so the `ScanPopup` acted just as before. 

when Ctrl+C+C is pressed, and failed to retrieve clipboard, `isTranslateWordFromClipboard` will be true, then when the clipboard change, the `handleTranslateWordFromClipboard()` will pop out once, then set `isTranslateWordFromClipboard` to false.

I think it works.

> the current code looks completely wrong.

What do you think is wrong with this solution.

**Comment by vedgy** [comment]
OK, maybe it works. But I feel the code is overcomplicated if all you want to achieve is delay retrieving the word from the clipboard. Have you tried my simple alternative fix? If it doesn't work, I am sure a simpler than in this PR solution (that requires at most one timer) is possible.

**Comment by nonwill** [comment]
I agree with @vedgy 's suggestion too.
The problems encountered with using the clipboard`Ctrl+C(+C)` to trigger a Popup searching depend on the specific application scenario. It is difficult to have a common and workable solution, so the first thing to do is to rule out as much as possible that it is other applications causing the negative impact(influence to each other), and if it must be resolved from GD's codes to reach an agreed outcome, the simpler the modification, the better. Because the method may not necessarily be universal.

**Comment by vedgy** [comment]
@wisherhxl, I am sorry for the misunderstanding. Now I understand your fix. But still I'd like to reduce its complexity. So instead of modifying downstream ScanPopup code, I suggest applying the workaround earlier.

> I guess maybe that's because when the popup is triggered, the other Qt program have not finished the setText procedure.

As I understand it, the problem is that GoldenDict handles `Ctrl+C+C` before the active application handles `Ctrl+C`. This issue only affects the hotkey, so let's apply a workaround in `MainWindow::hotKeyActivated()` to prevent regressions elsewhere. The simplest and safest possible workaround is to replace the `scanPopup->translateWordFromClipboard();` call with `QTimer::singleShot( 0, scanPopup, SLOT( translateWordFromClipboard() );`. If this doesn't help, binary-search for the lowest working delay to use in place of `0`, then increase this lowest delay slightly to increase the robustness of the workaround, but without increasing the risk of regressions much.

Since only GNU/Linux users report the Ctrl+C+C bug, it makes sense to limit the workaround to this platform: `#ifdef HAVE_X11`.

**Comment by vedgy** [comment]
Note that this workaround cannot possibly fix #1290 because GoldenDict's X11-specific hotkey code still won't work in Wayland.

**Comment by nonwill** [comment]
https://github.com/goldendict/goldendict/blob/f810c6bd724e61977b4e94ca2d8abfa5bd766379/mainwindow.cc#L3092
```
    connect( hotkeyWrapper.get(), SIGNAL( hotkeyActivated( int ) ),
             this, SLOT( hotKeyActivated( int ) ),
#ifdef Q_OS_WIN32
             hotkeyWrapper->handleViaDLL() ? Qt::QueuedConnection : Qt::AutoConnection );
#else
             Qt::AutoConnection );
#endif
  }
````

How about just to make hotKeyActivated being bounded with Qt::QueuedConnection? But if the SIGNAL is emited from a single thread(but not main), it's already been queued when the slot is triggered...

**Comment by vedgy** [comment]
> How about just to make hotKeyActivated being bounded with Qt::QueuedConnection?

One downside of this alternative workaround is that the Show/hide main window hotkey is affected, even though the issue does not apply to it.

**Comment by nonwill** [comment]
>Show/hide main window hotkey is affected

Yes, it's. 
I'm not sure what the real type is Qt::AutoConnection represents here on Linux. If hotkeyWrapper is running in a separate thread(type should be Qt::QueuedConnection at runtime), then it doesn't make any difference.

**Comment by wisherhxl** [comment]
I like the simplest solution of `QTimer::singleShot( 0, scanPopup, SLOT( translateWordFromClipboard() ) );` 
And 10ms works just fine on all my computers. 
So I create #1491 with this solution and close this one.
